ddraw/IDirect3DDeviceX.cpp

```diff
@@ -338,14 +338,95 @@ HRESULT m_IDirect3DDeviceX::SetTransform(D3DTRANSFORMSTATETYPE dtstTransformStat
 			break;
 		}
 
-		HRESULT hr = (*d3d9Device)->SetTransform(dtstTransformStateType, lpD3DMatrix);
-
-		if (SUCCEEDED(hr))
+		D3DMATRIX view;
+		if (Config.DdrawConvertHomogeneousW)
 		{
 #ifdef ENABLE_DEBUGOVERLAY
+			// Set the original matrix
 			DOverlay.SetTransform(dtstTransformStateType, lpD3DMatrix);
 #endif
+
+			if (dtstTransformStateType == D3DTS_VIEW)
+			{
+				D3DVIEWPORT9 Viewport9;
+				if (SUCCEEDED((*d3d9Device)->GetViewport(&Viewport9)))
+				{
+					const float width = (float)Viewport9.Width;
+					const float height = (float)Viewport9.Height;
+
+					// Replace the matrix with one that handles D3DFVF_XYZRHW geometry
+					ZeroMemory(&view, sizeof(D3DMATRIX));
+					view._11 = 2.0f / width;
+					view._22 = -2.0f / height;
+					view._33 = 1.0f;
+					view._41 = -1.0f;  // translate X
+					view._42 = 1.0f;   // translate Y
+					view._44 = 1.0f;
+
+					// Set flag
+					ConvertHomogeneous.IsTransformViewSet = true;
+
+					if (Config.DdrawConvertHomogeneousToWorld)
+					{
+						DirectX::XMVECTOR position, direction;
+						if (Config.DdrawConvertHomogeneousToWorldUseGameCamera)
+						{
+							// To reconstruct the 3D world, we need to know where the camera is and where it is looking
+							position = DirectX::XMVectorSet(lpD3DMatrix->_41, lpD3DMatrix->_42, lpD3DMatrix->_43, lpD3DMatrix->_44);
+							direction = DirectX::XMVectorSet(lpD3DMatrix->_31, lpD3DMatrix->_32, lpD3DMatrix->_33, lpD3DMatrix->_34);
+						}
+						else
+						{
+							position = DirectX::XMVectorSet(0.0f, 0.0f, 0.0f, 0.0f);
+							direction = DirectX::XMVectorSet(0.0f, 0.0f, 1.0f, 0.0f);
+						}
+
+						// Store the original matrix so it can be restored
+						ConvertHomogeneous.ToWorld_ViewMatrixOriginal = view;
```

Comment:
This line is not storing the original matrix.  Should this be changed to:

```c+
ConvertHomogeneous.ToWorld_ViewMatrixOriginal = *lpD3DMatrix;
```

Reply:
No because the original geometry still needs to be transformed so they work with the expected space.

The original view matrix is useless to transform geometry. It only exists to transform the lights.

This matrix is needed to make geometry look correctly when D3DFVF_XYZRHW is changed to D3DFVF_XYZW.

Reply:
Ok, I see.